### ISSUE TITLE
[java] ArrayIsStoredDirectly at enum constructor

### ISSUE THREAD

**Issue by frankegabor** [pmd/pmd#1413]
**Affects PMD Version:** tested on 6.3.0

**Rule:** [ArrayIsStoredDirectly](https://pmd.github.io/latest/pmd_rules_java_bestpractices.html#arrayisstoreddirectly)

**Description:**
See the code below. The array what is stored through the constructor is immutable, known at compile time, can not be changed runtime, I think at this case the array can be stored directly.

**Code Sample demonstrating the issue:**
```java
public enum DemoEnum {

    VAL1("str1", "str2"),
    VAL2("str3");

    private final String[] values;

    DemoEnum(String... values) {
        this.values = values;
    } 
}
```

**Running PMD through:** *[java]*

**Comment by adangel** [comment]
Thanks for the report.

In your example, you are right: the string array is known at compile time and is a compile time constant.

But this doesn't need to be the case in general, e.g. this is a valid program but the array can be changed at runtime:

```java
import java.util.Arrays;

public class ArrayIsStoredDirectlyTest {
    @org.junit.Test
    public void testEnum() {
        System.out.println(Test.A);
        System.out.println(Test.B);
        System.out.println(Test.C);

        strings[1] = "modified";
        System.out.println(Test.C);
    }

    public static String[] strings = new String[]{"a", "b"};

    public enum Test {
        A("a"),

        B("b".toUpperCase()),

        C(strings);

        private final String[] a;
        Test(String ... args) {
            a = args;
        }

        @Override
        public String toString() {
            return super.toString() + ":" + Arrays.toString(a);
        }
    }
}
```

This outputs the following:

```
A:[a]
B:[B]
C:[a, b]
C:[a, modified]
```

That's why we can't simply ignore enum constructors - we'd need to check additionally whether every enum constant is created with a compile time constant....

**Comment by adangel** [comment]
This is no longer reproducible under PMD 7.0.0.
